langchain/retrievers/perfect_retriever.py

```diff
@@ -0,0 +1,220 @@
+import json
+from abc import abstractmethod
+from typing import Any, Dict, List
+
+from langchain import LLMChain, PromptTemplate
+from langchain.agents import AgentExecutor
+from langchain.agents.agent_toolkits import VectorStoreInfo
+from langchain.llms import BaseLLM
+from langchain.schema import BaseOutputParser, BaseRetriever, Document
+from langchain.vectorstores.base import VectorStoreRetriever
+
+pinecone_format_instructions = """RESPONSE FORMAT
+----------------------------
+
+When responding to me please, please output a response in this format. Use a markdown code snippet formatted in the following schema:
+
+```json
+{{{{
+    "search_string": string \\ The string to compare to document contents,
+    "metadata_filter": {{{{
+        "<MetadataField>": {{{{
+            <Operator>: <Value>
+        }}}},
+        "<MetadataField>": {{{{
+            <Operator>: <Value>
+        }}}},
+    }}}} \\ The conditions on which to filter the metadata
+}}}}
+```
+
+If no metadata filters are needed you can leave "metadata_filter" as an empty map.
+
+Filtering metadata supports the following operators:
+    $eq - Equal to (number, string, boolean)
+    $ne - Not equal to (number, string, boolean)
+    $gt - Greater than (number)
+    $gte - Greater than or equal to (number)
+    $lt - Less than (number)
+    $lte - Less than or equal to (number)
+    $in - In array (string or number)
+    $nin - Not in array (string or number)"""
+
+pinecone_example = """EXAMPLE
+----------------------------
+
+Here is an example:
+
+DOCUMENT STORE DESCRIPTION: News headlines from around the world
+METADATA FIELDS: {{{{
+    "year": "The year the headline was published", 
```

Comment:
Do we need type information to be able to re-interpret the values provided in the filter as the appropriate type?

Reply:
ooh good idea, i like it